### PR TITLE
Fixes #27246 - Handle device identifiers in IPs

### PR DESCRIPTION
**Pull request by ekohl**
On Windows Facter can return an IPv6 link local including the device identifier:

    fe80::302f:51e0:1313:f50d%5

However, IPAddr doesn't handle this. This patch handles this case and cleans up the link local handling.

I could use some help about writing tests here.

### CODE REVIEW

**Comment by timogoebel** [comment]
> I could use some help about writing tests here.

@ekohl: I think you can just copy the test from https://github.com/theforeman/foreman/pull/4439 and change it to test for ink local addresses that include the device identifier.

**Comment by ekohl** [comment]
Updated with added tests which removed 2 bugs from the implementation.

File: app/models/host/base.rb [hunk 1]
```diff
@@ -371,6 +371,25 @@ def uptime_seconds
 
     private
 
+    def parse_ip_address(address, ignore_link_local = true)
```
Comment (timogoebel) [inline app/models/host/base.rb]: nit: `I'd make this a kwarg`.

Comment (ekohl) [reply app/models/host/base.rb]: That's like this, right?
```suggestion
    def parse_ip_address(address, ignore_link_local: true)
```

Comment (timogoebel) [reply app/models/host/base.rb]: Yep.

Comment (ekohl) [reply app/models/host/base.rb]: Updated

**Comment by timogoebel** [comment]
[test foreman]

**Comment by timogoebel** [comment]
[test foreman]

**Comment by timogoebel** [comment]
@ekohl: Do you mind rebasing this? Hopefully, this will unlock the tests.

**Comment by ekohl** [comment]
Rebased.

File: app/models/host/base.rb [hunk 2]
```diff
@@ -372,6 +372,25 @@ def uptime_seconds
...
+      rescue IPAddr::InvalidAddressError
+        if address.is_a?(String) && address.include?('%')
+          addr = IPAddr.new(address.split('%').first)
+        else
+          logger.warn "Ignoring invalid IP address #{address}"
+          return nil
+        end
+      end
+
+      (ignore_link_local && addr.link_local?) ? nil : addr.to_s
```
Comment (timogoebel) [inline app/models/host/base.rb]: `link_local?` is not available in all the ruby versions we support, that's why tests are failing.

Basically the method does this:
> Returns true if the ipaddr is a link-local address. IPv4 addresses in 169.254.0.0/16 reserved by RFC 3927 and Link-Local IPv6 Unicast Addresses in fe80::/10 reserved by RFC 4291 are considered link-local.

So you can replace it by `IPAddr.new('fe80::/10').include?(ip)` for IPv6 and `IPAddr.new('169.254.0.0/16').include?(ip)`.

```ruby
if ignore_link_local
  return if addr.ipv4? && IPAddr.new('169.254.0.0/16').include?(addr)
  return if addr.ipv6? && IPAddr.new('fe80::/10').include?(addr)
end

addr.to_s
```

**Review by timogoebel — CHANGES_REQUESTED** [review]

File: app/models/host/base.rb (hunk 2 above)
Comment (ekohl) [reply app/models/host/base.rb]: Of course, it was too easy. Looks like Ruby 2.5 introduced this so I'll update it.

Comment (ekohl) [reply app/models/host/base.rb]: Updated to first attempt the modern method. Once we drop Ruby 2.3 (1.25 I think) we can drop the else part.

**Review by timogoebel — APPROVED** [review]

**Comment by timogoebel** [comment]
Thanks, @ekohl.

**Pull request merged**